### PR TITLE
Fix AbstractOperator::recreate() for diamonds and diamond Task plans

### LINKED ISSUE #742: Fix Operator::recreate for diamonds
This looks nice, but is... inefficient:

screen shot 2018-02-27 at 16 04 57

 I would expect a diamond shape.
To avoid confusion: The screenshot says query 6 because it is zero-indexed.

### PR DESCRIPTION
**Pull request by mrzzzrm**
Fix #742 

 - Also fixes `OperatorTask::make_tasks_from_operator()` for diamond shapes
 - Most of the changes are just the Operators going from `recreate(args)` to `_on_recreate(args, left, right)` and can be skipped during review. Look at AbstractOperator, OperatorTask::make_tasks_from_operators() and the Tests.
 - Use similar pattern to LQPs to recreate PQPs
 - Add `_on_recreate()` to operators that still lacked it
 - Remove SQLQueryOperator (again?!)

### CODE REVIEW

File: src/lib/operators/abstract_operator.cpp [hunk 1]
```diff
@@ -19,6 +19,8 @@ AbstractOperator::AbstractOperator(const std::shared_ptr<const AbstractOperator>
     : _input_left(left), _input_right(right) {}
 
 void AbstractOperator::execute() {
+  DebugAssert(!_output, "Operator has already been executed");
```
Comment (mrks) [inline src/lib/operators/abstract_operator.cpp]: A great idea!

File: src/lib/scheduler/operator_task.cpp [hunk 2]
```diff
@@ -23,33 +23,32 @@ std::string OperatorTask::description() const {
...
-    op->set_operator_task(task);
-    add_task = true;
-  }
+    std::shared_ptr<AbstractOperator> op, std::vector<std::shared_ptr<OperatorTask>>& tasks,
+    std::unordered_map<std::shared_ptr<AbstractOperator>, std::shared_ptr<OperatorTask>>& task_by_op) {
+  const auto task_by_op_it = task_by_op.find(op);
+  if (task_by_op_it != task_by_op.end()) return task_by_op_it->second;
+
+  const auto task = std::make_shared<OperatorTask>(op);
+  task_by_op.emplace(op, task);
```
Comment (mrks) [inline src/lib/scheduler/operator_task.cpp]: Why does this become necessary? Adding this map adds some lookup costs - couldn't we go by the `operator_task()` that we had before and simply clear it when we recreate an operator?

File: src/test/tasks/operator_task_test.cpp [hunk 3]
```diff
@@ -64,4 +65,35 @@ TEST_F(OperatorTaskTest, DoubleDependencyTasksFromOperatorTest) {
   auto expected_result = load_table("src/test/tables/joinoperators/int_inner_join.tbl", 2);
   EXPECT_TABLE_EQ_UNORDERED(expected_result, tasks.back()->get_operator()->get_output());
 }
+TEST_F(OperatorTaskTest, MakeDiamonShape) {
```
Comment (mrks) [inline src/test/tasks/operator_task_test.cpp]: *Diamond*

File: src/lib/scheduler/operator_task.cpp (hunk 2 above)
Comment (mrzzzrm) [reply src/lib/scheduler/operator_task.cpp]: Having the Operator know that there is such a concept as the OperatorTask always seemed dirty to me. Generally, objects with knowledge about their enclosing object are anti patterns. This even looks like a circular reference, now that I look at it. (which we could of course work around) (Humongous amounts of memory, anyone? https://github.com/hyrise/hyrise/issues/656)

Your call, but I'd like to keep it this way.

Comment (mrks) [reply src/lib/scheduler/operator_task.cpp]: Yes, this is circular, but could easily be broken with a `weak_ptr`.

I believe that there was another reason why we introduced it, which was to make sure that no operator was scheduled twice: 92058cf02e67ef42fc053342b9637097cf7ac67e

**Comment by mrzzzrm** [comment]
@mrks if everything for this is fine later, merge without my approval, probably won't look at this again today.

**Pull request merged**